lib/flex_algo.h

```diff
@@ -0,0 +1,172 @@
+
+	/*
+	 * This property can be freely extended among different routing
+	 * protocols. Since Flex-Algo is an IGP protocol agnostic, both IS-IS
+	 * and OSPF can implement Flex-Algo. The struct flex_algo thus provides
+	 * the general data structure of Flex-Algo, and the value of extending
+	 * it with the IGP protocol is provided by this property.
+	 */
+	void *data;
+};
+
+typedef void *(*flex_algo_allocator_t)(void *);
+typedef void (*flex_algo_releaser_t)(void *);
+
+struct flex_algos {
+	flex_algo_allocator_t allocator;
+	flex_algo_releaser_t releaser;
+	struct list *flex_algos;
+};
+
+/*
+ * Affinity Map Utilities
+ */
+struct affinity_maps *affinity_maps_alloc(void);
+void affinity_maps_free(struct affinity_maps *maps);
+void affinity_map_set(struct affinity_maps *maps, const char *name, int pos);
+void affinity_map_unset(struct affinity_maps *maps, const char *name);
+struct affinity_map *affinity_map_get(const struct affinity_maps *maps,
+				      const char *name);
+char *affinity_map_name_get(const struct affinity_maps *maps, int pos);
+
+/*
+ * Admin Group Utilities
+ */
+char *admin_group_string(char *out, size_t sz, int indent,
+			 const struct admin_group *ag);
+char *admin_group_standard_print(char *out, int indent, uint32_t bitmap);
+char *admin_group_print(char *out, int indent, const struct admin_group *ag);
+void admin_group_copy(struct admin_group *dst, const struct admin_group *src);
+void admin_group_init(struct admin_group *ag);
+void admin_group_term(struct admin_group *ag);
+uint32_t admin_group_get_offset(const struct admin_group *ag,
+				size_t oct_offset);
+void admin_group_set(struct admin_group *ag, size_t pos);
+void admin_group_unset(struct admin_group *ag, size_t pos);
+int admin_group_get(const struct admin_group *ag, size_t pos);
+void admin_group_set_offset(struct admin_group *ag, size_t pos,
+			    size_t oct_offset);
+void admin_group_unset_offset(struct admin_group *ag, size_t pos,
+			      size_t oct_offset);
+void admin_group_bulk_set(struct admin_group *ag, uint32_t bitmap,
+			  size_t oct_offset);
+size_t admin_group_size(const struct admin_group *ag);
+size_t admin_group_nb_words(const struct admin_group *ag);
+void admin_group_clear(struct admin_group *ag);
+bool admin_group_zero(const struct admin_group *ag);
+bool admin_group_explicit_zero(const struct admin_group *ag);
+void admin_group_allow_explicit_zero(struct admin_group *ag);
+void admin_group_disallow_explicit_zero(struct admin_group *ag);
+
+bool admin_group_match_any(const struct admin_group *fad_ag,
+			   const uint32_t *link_std_ag,
+			   const struct admin_group *link_ag);
+bool admin_group_match_all(const struct admin_group *fad_ag,
+			   const uint32_t *link_std_ag,
+			   const struct admin_group *link_ag);
+
+/*
+ * Flex-Algo Utilities
+ */
+struct flex_algos *flex_algos_alloc(flex_algo_allocator_t allocator,
+				    flex_algo_releaser_t releaser);
+struct flex_algo *flex_algo_alloc(struct flex_algos *flex_algos,
```

Comment:
Again don't see a need for this? This should be well managed within the library itself. Clients don't need to manage allocations for the data managed by the library.